### ISSUE TITLE
Feature request: additional logging to debug connection pool deadlocks

### ISSUE THREAD

**Issue by xxgreg** [brettwooldridge/HikariCP#442]
I ran into a problem caused by code in my application which would request and hold on to two connections simultaneously on the same thread. When testing the system under heavy load with 50 threads sharing a pool of 10 connections it would deadlock. i.e. All of the application threads where blocked in getConnection(), and no threads were waiting on the database to complete work.

To debug this I added logging when a thread called getConnection() but already held an open connection It logs two stacktraces, one for the initial call, and one for the subsequent call.

It would be nice to have a feature like this added to Hikari.

Here is the quick hack I used to debug my application as an example of what I mean:

https://github.com/xxgreg/HikariCP/commit/bef009ddf775e1d8473a285d06335903234e4400

**Comment by brettwooldridge** [comment]
@xxgreg I do appreciate the effort required to diagnose an issue like this.  Having said that, providing these kinds of checks in the "hot path" of ``getConnection()`` would crush performance.

While in the case of your application opening and holding two connections in the same thread resulting in an issue, there are many frameworks, including Hibernate and Play/Slick that do so routinely as a matter of course.  Unfortunately, I don't think the diagnostic code is universally applicable.

Of course the nice thing about open source is that you were able to instrument HikariCP to suit your needs and diagnose the issue.

**Comment by xxgreg** [comment]
If an application is relying on being able to obtain multiple connections simultaneously from the same pool how can it protect itself against a deadlock? This sounds like a recipe for trouble to me - given sufficient contention it will always be possible to have all threads waiting on each other to release a connection in order to make progress.

**Comment by brettwooldridge** [comment]
I'm not sure how your application in particular got into that pickle, but the calculation of pool size in this scenario in order to avoid deadlock is fairly simple:

*pool size* = t<sup>n</sup> x (c<sup>m</sup> - 1) + 1

Where t<sup>n</sup> is the maximum number of threads, and c<sup>m</sup> is the maximum number of simultaneous connections required by a single thread.

For example, imagine three threads (t<sup>n</sup>=3), each of which requires four connections to perform some task (c<sup>m</sup>=4).  The pool size required to ensure that deadlock is never possible is:

*pool size* = 3 x (4 - 1) + 1 = 8

If you have a maximum of ten threads (t<sup>n</sup>=10), each of which requires two connections to perform some task (c<sup>m</sup>=4).  The pool size required to ensure that deadlock is never possible is:

*pool size* = 10 x (2 - 1) + 1 = 11

**Comment by xxgreg** [comment]
Thanks for the detailed comments.

The app had 50 jetty threads, 10 connections in the pool, connections were acquired and released multiple times per request, and two simultaneous connections were required for a thread to complete (this was introduced accidentally). So it was a deadlock waiting to happen. Glad that we caught it during load testing and not in production.

I wonder whether it is worth adding what you have written above to your very helpful article about connection pool sizing. Perhaps some users of Hibernate and Slick are unaware that their connection pool needs to be larger than their number of application threads to avoid deadlock. 

Currently there is this:

"The prospect of "pool-locking" has been raised with respect to single actors that acquire many connections. This is largely an application-level issue. Yes, increasing the pool size can alleviate lockups in these scenarios, but we would urge you to examine first what can be done at the application level."

**Comment by brettwooldridge** [comment]
@xxgreg Good idea, I will add the above to the pooling page, as well as the FAQ.
